Development environment guide

christopher-hakkaart opened the pull request:
Added a new guide for creating development environment page. Added tabs to help split Windows/macOS/Linux.

Development environment
- VS Code
- Extensions
    - Nextflow
    - nf-core
    - Remote development
- Docker
- Git
- WSL

**Updates:**
- Moved from **Getting started** to **Tutorials**
- **Dev containers** section has been removed

@netlify /developer-env

ewels commented on docs/developer-env.md:
`````diff
@@ -0,0 +1,473 @@
+
+See {ref}`vs-code-page` for more information about the Nextflow extension.
+
+````{tabs}
`````
This one is the same for all OS's right? No need for tabs here?

christopher-hakkaart replied:
Changed locally. Will push shortly.

ewels commented on docs/developer-env.md:
`````diff
@@ -0,0 +1,473 @@
+:::
+
+
+````{tabs}
`````
Also the same for all OSs

christopher-hakkaart replied:
Changed locally. Will push shortly.

ewels commented on docs/developer-env.md:
`````diff
@@ -0,0 +1,473 @@
+
+````
+
+## Extensions
`````
Worth mentioning the nf-core extension pack for other goodies?

ewels replied:
Should be an h3 nested under VSCode? Or be called "VS Code Extensions"

christopher-hakkaart replied:
Nice. I'll add a section for nf-core and I'll update the title to "VS Code Extensions" to improve clarity and prevent too much nesting.

ewels commented on docs/developer-env.md:
````diff
@@ -0,0 +1,473 @@
+
+```{group-tab} macOS
+
+Git installed is already installed on new versions of macOS. You can activate it through the terminal running `git version`. If Git is not installed, you can install the latest version of Git using several methods:
````
> Git installed is already installed on new versions of macOS.

You sure about that? Pretty sure that it hasn't shipped as default for a while now, you get prompted to install the XCode developer pack if you try to use it.

https://forums.developer.apple.com/forums/thread/672087

> DTS Engineer
> Apple
> Jan ’21
> `git` is not installed on macOS by default. The binary at `/usr/bin/git` is a shim that redirects any calls to that command line tool to the version in the current version of Xcode or the Command Line Tools downloadable package. This is so the tool is updated as Xcode updates, among other benefits.

christopher-hakkaart replied:
I'll double-check. The article I read said it was but it might have been out of date.

christopher-hakkaart replied:
Updated

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,473 @@
+Setting up a Nextflow development environment is a prerequisite for creating, testing, and optimizing data analysis pipelines. The steps below outline recommended tools for setting up an optimal Nextflow development environment.
+
+:::{note}
+If you are using a Windows computer, you first need to install and configure the Windows Subsystem for Linux (WSL). See {ref}`wsl` for installation instructions.
```
Why not put this at the top of the document then?

christopher-hakkaart replied:
Windows is used by a minor fraction of users, and if you're in bioinformatics, you likely already have this installed/setup. I wanted to keep it broad before getting specific for Windows users only.

kenibrewer replied:
I think I like Chris's approach here too.

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,473 @@
+
+# Environment setup
+
+Setting up a Nextflow development environment is a prerequisite for creating, testing, and optimizing data analysis pipelines. The steps below outline recommended tools for setting up an optimal Nextflow development environment.
```
A brief TOC here would help I think, not sure if that's non-standard for these docs though.

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,473 @@
+(developer-env)=
+
+# Environment setup
```
I think it'd be good to have an admonition / note thing saying "this is once you've installed Nextflow, click here for instructions on that". So that if anyone drops directly onto this page they're not confused.

christopher-hakkaart replied:
Added extra line.

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,473 @@
+
+## Development Containers
+
+Development Containers (Dev Containers), are Docker containers that are specifically configured to provide a fully featured development environment. They can be used to run an application, to separate tools, libraries, or runtimes needed for working with a codebase. Dev Containers can be run locally or remotely, in a private or public cloud, and in a variety of supporting tools and editors.
```
Maybe link to https://containers.dev/

christopher-hakkaart replied:
Added

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,473 @@
+1. Enter **Dev Containers: Reopen in Container** in the VS Code Command Palette and reopen your project. You should now see the name of the container ("Nextflow Dev Container" if using the above example) in the bottom left corner of VS Code.
+
+:::{note}
+Dev Containers can also be used by GitHub Codespaces in VS Code or the browser. See [GitHub Codespaces](https://code.visualstudio.com/docs/remote/codespaces) for more information.
```
Soon also GitPod (Flex)

christopher-hakkaart replied:
Added

ewels reviewed (commented):
Very nice!

pditommaso commented on docs/developer-env.md:
````diff
@@ -0,0 +1,473 @@
+- Set runtime arguments.
+- Reuse or extend your existing Docker Compose setup.
+- Add more Advanced container configuration.
+
+To create a dev container with an existing image:
+
+1. Create `.devcontainer/devcontainer.json` in the root of your project.
+1. Add an image with the Nextflow tooling and VS Code customizations to the `devcontainer.json`. For example:
+
+    ```json
+    {
+        "name": "Nextflow Dev Container",
+        "image": "nfcore/gitpod:latest",
+        "remoteUser": "vscode",
+        "runArgs": ["--privileged"],
+
+        // Configure tool-specific properties.
+        "customizations": {
+            // Configure properties specific to VS Code.
+            "vscode": {
+                // Set *default* container specific settings.json values on container create.
+                "settings": {
+                    "python.defaultInterpreterPath": "/opt/conda/bin/python"
+                },
+
+                // Add the IDs of extensions you want installed when the container is created.
+                "extensions": ["ms-python.python", "ms-python.vscode-pylance", "nf-core.nf-core-extensionpack", "nextflow.nextflow"]
+            }
+        }
+    }
+    ```
+
+    :::{note}
+    Instead of using a prebuilt image, a custom Dockerfile may also live in the `.devcontainer` folder. You can replace the image property in `devcontainer.json` with dockerfile and utilize the custom container. See [Create a Dev Container](https://code.visualstudio.com/docs/devcontainers/create-dev-container) for more information.
+    :::
+
+1. Enter **Dev Containers: Reopen in Container** in the VS Code Command Palette and reopen your project. You should now see the name of the container ("Nextflow Dev Container" if using the above example) in the bottom left corner of VS Code.
+
+:::{note}
+Dev Containers can also be used by GitHub Codespaces in VS Code or the browser. See [GitHub Codespaces](https://code.visualstudio.com/docs/remote/codespaces) for more information.
````
not sure this makes much sense in the context of nextflow workflow in which each process can have its own container. Our best practice is Conda + Wave for container reproducibility and portability

christopher-hakkaart replied:
Fair point. I'm not committed to keeping this. It's very thin on details and can get complicated quickly. I'll comment it out.

ewels replied:
Yeah agree with this. I think that devcontainers for the pipeline development environment is not incompatible with using conda / wave etc for actually running the pipeline.

pditommaso commented:
Not sure this is material for nextflow documentation, this is more material for tutorial or blog post

kenibrewer commented:
> Not sure this is material for nextflow documentation, this is more material for tutorial or blog post

I've had four people in the last week (1 at summit and 3 at ASHG) tell me that they had a hard time figuring out how to install Nextflow from the documentation. Two of them gave up and quit and never used Nextflow, one of them needed personal help from me, and one was running an ancient version of nextflow that came pre-installed as a workaround. 

Those user experiences make me think this is a very important addition to the main docs. I also think it would be better featured prominently as one of the first pages new users encounter on the docs page. A separate tutorial page or blog post will be harder to find and won't get continued ongoing attention to make sure we keep it up to date.

pditommaso commented:
> I've had four people in the last week  .. 

Understand and it makes sense. However, I'm contrary having dev containers in this context. Provided it could help, it should not make part of the best practices we are promoting for container usage both for dev and production deployment

justinegeffen commented on docs/developer-env.md:
```diff
@@ -0,0 +1,530 @@
+1. Double-click `Docker.dmg` to open the installer.
+1. Drag the Docker icon to the **Applications** folder to make it available in the macOS Launchpad.
+1. Double-click **Docker.app** in the **Applications** folder to start Docker.
+1. Review the Docker Subscription Service Agreement and, if you agree, select **Accept** to continue.
```
I think this note should be added here too

:::{note}
    Docker Desktop won't run if you do not agree to the terms. You can choose to accept the terms at a later date by opening Docker Desktop.
    :::

justinegeffen replied:
That said, if I have to accept the terms for Docker Desktop to run but I can also accept them at a later date does that mean Docker Desktop will work anyway? 

Perhaps we can remove this note throughout and rather have something like: 

"Review the Docker Subscription Service Agreement and select **Accept** to continue. Docker Desktop won't run if you do not agree to the terms."

Obviously this depends on the exact behavior when someone does/doesn't accept the terms which I'm not 100% sure of.

justinegeffen commented on docs/developer-env.md:
```diff
@@ -0,0 +1,530 @@
+To install Git on macOS with [Xcode](https://developer.apple.com/xcode/):
+
+1. Open the App Store on your Mac.
+1. Sign in to your Apple Account.
```
Technically the user should already be signed in but if not, I'd move this to step 1.

christopher-hakkaart replied:
Manually removed

justinegeffen commented on docs/developer-env.md:
```diff
@@ -0,0 +1,530 @@
+
+1. In the Windows search bar, enter 'features' to bring up the **Turn Windows Features on and off** dialog.
+1. Scroll down and check **Windows Subsystem for Linux**.
+1. Select **OK** and restart Windows.
```
```suggestion
1. Select **OK**.
```

justinegeffen replied:
Is the **OK** a confirmation to restart Windows or is there a next step?

justinegeffen commented on docs/developer-env.md:
```diff
@@ -0,0 +1,530 @@
+
+To enable WSL on Windows using Powershell or Windows Command Prompt:
+
+1. Open PowerShell or Windows Command Prompt in administrator mode by right-clicking and selecting **Run as administrator**.
```
```suggestion
1. Right-click and select **Run as administrator** to use PowerShell or Windows Command Prompt in administrator mode.
```

justinegeffen commented on docs/developer-env.md:
```diff
@@ -0,0 +1,530 @@
+    :::
+
+1. When prompted, restart Windows.
+1. After restarting Windows, open the Ubuntu distribution using the **Start** menu and create a new Linux **User Name** and **Password** when prompted.
```
```suggestion
1. After restarting Windows, open the Ubuntu distribution and create a new Linux **User Name** and **Password** when prompted.
```

justinegeffen commented on docs/developer-env.md:
```diff
@@ -0,0 +1,530 @@
+
+## Development Containers
+
+[Development Containers](https://containers.dev/) (Dev Containers), are Docker containers that are specifically configured to provide a fully featured development environment. They can be used to run an application, to separate tools, libraries, or runtimes needed for working with a codebase. Dev Containers can be run locally or remotely, in a private or public cloud, and in a variety of supporti ...
```
```suggestion
[Development Containers](https://containers.dev/) (Dev Containers), are Docker containers that are specifically configured to provide a full-featured development environment. They can be used to run an application, to separate tools, libraries, or runtimes needed for working with a codebase. Dev Containers can be run locally or remotely, in a private or public cloud, and in a variety of supporting tools and editors.
```

justinegeffen commented on docs/developer-env.md:
```diff
@@ -0,0 +1,530 @@
+
+### Development Containers extension
+
+The VS Code Dev Containers extension lets you use a container as a full-featured development environment. It allows you to open any folder inside (or mounted into) a container and take advantage of VS Code's full feature set.
```
We've said "full-featured" 3 times in the past few paragraphs. Maybe we can remove these two or use a different adjective? I'm coming up blank right now though!

justinegeffen reviewed (commented):
This is awesome! 

Please don't be alarmed at the number of comments - the content is great - just a few editorial tweaks here and there. Happy to workshop the feedback if it's unclear. :)

christopher-hakkaart commented on docs/developer-env.md:
````diff
@@ -0,0 +1,530 @@
+1. Search for **Nextflow**.
+1. Select **Install**.
+
+```
````
It's the end of the group tab

christopher-hakkaart commented on docs/developer-env.md:
```diff
@@ -0,0 +1,530 @@
+
+## Docker
+
+Containerization enables the creation of self-contained and fully reproducible computational pipelines by bundling a script's binary dependencies into a standardized and portable format. Containers can be executed on any platform that supports a container runtime and ensures consistency across different environments.
```
Nicer progression 👍 Manually edited

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,534 @@
+
+The nf-core extension pack includes 12 extensions:
+
+- [Apptainer/Singularity](https://marketplace.visualstudio.com/items?itemName=onnovalkering.vscode-singularity) - Provides syntax highlighting for Apptainer/Singularity definition files.
+- [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) - A basic spell checker that works well with camelCase code.
+- [Docker](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker) - Makes it easy to create, manage, and debug containerized applications.
+- [EditorConfig](https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig) - Support for EditorConfig project files for code standardisation.
+- [indent-rainbow](https://marketplace.visualstudio.com/items?itemName=oderwat.indent-rainbow) - Highlight indentation levels, especially if they're inconsistent.
+- [Nextflow](https://marketplace.visualstudio.com/items?itemName=nextflow.nextflow) - Nextflow language support.
+- [Prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) - Code formatter using Prettier.
+- [Rainbow CSV](https://marketplace.visualstudio.com/items?itemName=mechatroner.rainbow-csv) - Highlight columns in .CSV files in different colors.
+- [Ruff](https://marketplace.visualstudio.com/items?itemName=charliermarsh.ruff) - An extremely fast Python linter and code formatter, written in Rust.
+- [Todo Tree](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree) - Show TODO, FIXME, etc. comment tags in a tree view.
+- [YAML](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) - YAML Language Support by Red Hat, with built-in Kubernetes syntax support.
+- [Markdown Extended](https://marketplace.visualstudio.com/items?itemName=jebbs.markdown-extended) - Gives Markdown previews, including admonitions.
```
I'm slightly wary of actually listing them out like this, as it's the kind of thing that could get stale quite quickly. Would prefer something more generic like "a number of useful extensions that come recommended from community members, such as ..." then like 2 or 3 specific examples.

christopher-hakkaart replied:
I can see your point. I'll shrink this (and the other extensions) down

christopher-hakkaart replied:
I've shorted this text and removed the repeated code blocks.

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,534 @@
+- Parameter schemas
+- DAG previews
+
+See {ref}`vs-code-page` for more information about the Nextflow extension.
```
```suggestion
See the {ref}`vs-code-page` documentation for more information about the Nextflow extension.
```

ewels commented on docs/developer-env.md:
`````diff
@@ -0,0 +1,534 @@
+
+See {ref}`vs-code-page` for more information about the Nextflow extension.
+
+````{tabs}
`````
Can we have just one of these for "installing VSCode extensions"? Feels bulky to have essentially the identical content repeated for each set of VSCode extensions.

At the moment the VSCode section is like, most of the page. I think making each subsection shorter would help.

christopher-hakkaart replied:
Done.

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,534 @@
+
+- [Remote - SSH](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh) - Work with source code in any location by opening folders on a remote machine/VM using SSH.
+- [Remote - Tunnels](https://marketplace.visualstudio.com/items?itemName=ms-vscode.remote-server) - Work with source code in any location by opening folders on a remote machine/VM using a VS Code Tunnel (rather than SSH).
+- [Dev Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) - Work with a separate toolchain or container based application by opening any folder mounted into or inside a container.
+- [WSL](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl) - Get a Linux-powered development experience from the comfort of Windows by opening any folder in the Windows Subsystem for Linux.
```
I'm a bit confused by these as they're not really remote..? Maybe worth addressing this given the heading?

christopher-hakkaart replied:
I think they are treated as remote file systems. Using the same language as used in the pack description. Main purpose is for windows developers and those accessing remote file systems.

bentsherman replied:
The VSCode experience in WSL is basically the same as native Windows

Also, this text seems overly promotional

christopher-hakkaart replied:
I agree the text is very promotional. I took the text from [here](https://code.visualstudio.com/docs/remote/remote-overview), I'll tone it down.

ewels commented on docs/developer-env.md:
````diff
@@ -0,0 +1,534 @@
+
+```{group-tab} macOS
+
+To install Git on macOS with [Homebrew](https://docs.brew.sh/):
````
```suggestion
Arguably the easiest way to install Git on macOS is with [Homebrew](https://docs.brew.sh/):
```

christopher-hakkaart replied:
I'd rather keep the language neutral but I can see why having two options is confusing. I'll drop Xcode and point to external docs.

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,534 @@
+
+1. Once complete, run `git version` to verify Git was installed.
+
+To install Git on macOS with [Xcode](https://developer.apple.com/xcode/):
```
```suggestion
Alternatively, you can install Git on macOS as part of [Xcode](https://developer.apple.com/xcode/):
```

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,534 @@
+
+(wsl)=
+
+## Windows Subsystem for Linux
```
This still feels to me to be more part of the Nextflow installation, rather than Dev Environment.

Personally I'd rather remove it from this page entirely.

ewels commented on docs/developer-env.md:
```diff
@@ -0,0 +1,534 @@
+
+See [Set up a WSL development environment](https://learn.microsoft.com/en-us/windows/wsl/setup/environment) for more information.
+
+## Development Containers
```
I wonder if this section would be better off as a blog post that we can point to in an admonition from this page.

(a) This part feels quite a lot more detailed than the rest of the page and (b) I feel that devcontainers are more of a "this can be a nice thing" rather than a requirement. I, and I guess the majority, of Nextflow developers have never used this.

christopher-hakkaart replied:
I think it's emerging but not especially common. I'll drop it from this page and add it to the list of possible blog posts (maybe something @kenibrewer would want to write)

bentsherman replied:
We also have a "Tutorials" section in the Nextflow docs as a temporary home for tutorial-like content until we have something better on seqera.io, so if you want something more permanent than a blog post you can put it there.

Blog post or tutorial would be more appropriate for dev containers for now IMO

christopher-hakkaart replied:
Okay, great. I think it's agreed that it should be removed from the PR. I'll keep the text in case we can recycle it for a blog or tutorial. Depending on how it's pitched, it could be wrapped up into a tutorial or blog. Either way, we can put it aside for now.

christopher-hakkaart replied:
I've commented this out for now. I'll keep the text incase someone wants to write a blog in the new year.

ewels reviewed (commented):
Content is looking great, though I think that the VSCode part can be condensed a bit to help with general readability and maybe yank the devcontainers into a blog post. Otherwise, comments are very minor.

Great work! 👏🏻

christopher-hakkaart commented:
I've removed the dev containers section, removed repeating text boxes, removed bullet point examples, and move the whole page down to the tutorials section.

pditommaso commented:
@christopher-hakkaart sorry, didn't realised my comment had been addressed 👍
